File: lib/crypto.js
import crypto from 'node:crypto'
import fs from 'node:fs/promises'
import path from 'node:path'
import zlib from 'node:zlib'
import { promisify } from 'node:util'
import { fromBase64Url, toBase64Url } from './base64url.js'

const gzipAsync = promisify(zlib.gzip)
const gunzipAsync = promisify(zlib.gunzip)

export const ENCRYPTED_PART_SIZE = 32 * 1024 * 1024
export const PLAIN_CHUNK_SIZE = ENCRYPTED_PART_SIZE - 16
export const DIRECT_UPLOAD_LIMIT_BYTES = 100 * 1024 * 1024
export const MULTIPART_THRESHOLD_BYTES = 10 * 1024 * 1024

export function buildChunkIv(noncePrefix, chunkIndex) {
  const iv = Buffer.alloc(12)
  noncePrefix.copy(iv, 0, 0, 8)
  iv.writeUInt32BE(chunkIndex, 8)
  return iv
}

export function resolveCompressionMode() {
  return 'gzip'
}

export function usesChunkCompression(meta) {
  return meta?.compression?.algorithm === 'gzip'
}

export function normalizeCompressionMeta(rawCompression) {
  if (!rawCompression || typeof rawCompression !== 'object') {
    return { algorithm: 'none', scope: 'chunk' }
  }

  const algorithm = String(rawCompression.algorithm || 'none').toLowerCase()
  if (algorithm === 'gzip') {
    return { algorithm: 'gzip', scope: 'chunk' }
  }

  return { algorithm: 'none', scope: 'chunk' }
}

export async function compressPlainChunk(plainChunk) {
  return gzipAsync(plainChunk)
}

export async function decompressPlainChunk(compressedChunk) {
  return gunzipAsync(compressedChunk)
}

export function encryptChunk(payloadChunk, key, noncePrefix, chunkIndex) {
  const iv = buildChunkIv(noncePrefix, chunkIndex)
  const cipher = crypto.createCipheriv('aes-256-gcm', key, iv)
  const encrypted = Buffer.concat([cipher.update(payloadChunk), cipher.final()])
  const tag = cipher.getAuthTag()
  return Buffer.concat([encrypted, tag])
}

export function buildClientEncryptionMeta({
  plainChunkSize,
  chunkCount,
  noncePrefix,
  originalSize,
  mimeType,
}) {
  return {
    version: 2,
    algorithm: 'AES-GCM',
    compression: { algorithm: 'gzip', scope: 'chunk' },
    chunkSize: plainChunkSize,
    chunkCount,
    noncePrefix: toBase64Url(noncePrefix),
    originalSize,
    mimeType: mimeType || 'application/octet-stream',
  }
}

export function calculateEncryptedSize(originalSize, plainChunkSize = PLAIN_CHUNK_SIZE) {
  const chunkCount = Math.ceil(originalSize / plainChunkSize) || 1
  let encryptedSize = 0

  for (let chunkIndex = 0; chunkIndex < chunkCount; chunkIndex += 1) {
    const start = chunkIndex * plainChunkSize
    const plainSize = Math.min(originalSize - start, plainChunkSize)
    encryptedSize += plainSize + 16
  }

  return { chunkCount, encryptedSize }
}

export function createEncryptionContext({
  originalSize,
  originalName = 'upload',
  mimeType = 'application/octet-stream',
  tokenBytes = crypto.randomBytes(32),
  noncePrefix = crypto.randomBytes(8),
  plainChunkSize = PLAIN_CHUNK_SIZE,
}) {
  if (tokenBytes.length !== 32) throw new Error('Encryption token must be 32 bytes.')
  if (noncePrefix.length !== 8) throw new Error('Nonce prefix must be 8 bytes.')

  const { chunkCount, encryptedSize } = calculateEncryptedSize(originalSize, plainChunkSize)
  const meta = buildClientEncryptionMeta({
    plainChunkSize,
    chunkCount,
    noncePrefix,
    originalSize,
    mimeType,
  })

  return {
    token: toBase64Url(tokenBytes),
    tokenBytes,
    noncePrefix,
    meta,
    chunkCount,
    encryptedSize,
    originalName,
    mimeType,
    originalSize,
    plainChunkSize,
    compression: 'gzip',
  }
}

export async function resolveUploadFile(filePath, options = {}) {
  const absolutePath = path.resolve(filePath)
  const stat = await fs.stat(absolutePath)

  if (!stat.isFile()) {
    throw new Error('Upload path must be a file.')
  }

  const originalName = options.originalName || path.basename(absolutePath)
  const mimeType = options.mimeType || guessMimeType(originalName)

  return {
    absolutePath,
    context: createEncryptionContext({
      originalSize: stat.size,
      originalName,
      mimeType,
    }),
  }
}

async function readPlainChunk(fileHandle, chunkIndex, originalSize, plainChunkSize = PLAIN_CHUNK_SIZE) {
  const start = chunkIndex * plainChunkSize
  const length = Math.min(plainChunkSize, originalSize - start)
  const buffer = Buffer.alloc(length)
  const { bytesRead } = await fileHandle.read(buffer, 0, length, start)

  if (bytesRead !== length) {
    throw new Error('File ended before the upload was complete.')
  }

  return buffer
}

export async function encryptChunkFromFile(fileHandle, chunkIndex, context) {
  const plainChunk = await readPlainChunk(
    fileHandle,
    chunkIndex,
    context.originalSize,
    context.plainChunkSize,
  )

  let payload = await compressPlainChunk(plainChunk)

  return encryptChunk(payload, context.tokenBytes, context.noncePrefix, chunkIndex)
}

export function decryptChunk(encryptedChunk, tokenBytes, noncePrefix, chunkIndex) {
  if (encryptedChunk.length < 16) {
    throw new Error('Encrypted chunk is too small.')
  }

  const iv = buildChunkIv(noncePrefix, chunkIndex)
  const ciphertext = encryptedChunk.subarray(0, encryptedChunk.length - 16)
  const tag = encryptedChunk.subarray(encryptedChunk.length - 16)
  const decipher = crypto.createDecipheriv('aes-256-gcm', tokenBytes, iv)
  decipher.setAuthTag(tag)
  return Buffer.concat([decipher.update(ciphertext), decipher.final()])
}

export async function decryptPlainChunkFromEncrypted(encryptedChunk, tokenBytes, noncePrefix, chunkIndex, meta) {
  const payload = decryptChunk(encryptedChunk, tokenBytes, noncePrefix, chunkIndex)

  if (!usesChunkCompression(meta)) {
    return payload
  }

  return decompressPlainChunk(payload)
}

export function importToken(token) {
  const tokenBytes = fromBase64Url(token)
  if (tokenBytes.length !== 32) {
    throw new Error('Invalid encryption token length.')
  }
  return tokenBytes
}

export function normalizeClientEncryptionMeta(rawMeta) {
  if (!rawMeta || typeof rawMeta !== 'object') return null

  const chunkSize = Number(rawMeta.chunkSize)
  const chunkCount = Number(rawMeta.chunkCount)
  const originalSize = Number(rawMeta.originalSize)

  if (!Number.isFinite(chunkSize) || chunkSize <= 0) return null
  if (!Number.isFinite(chunkCount) || chunkCount <= 0) return null
  if (!Number.isFinite(originalSize) || originalSize < 0) return null
  if (typeof rawMeta.noncePrefix !== 'string' || !rawMeta.noncePrefix) return null

  return {
    version: Number(rawMeta.version) || 1,
    algorithm: String(rawMeta.algorithm || 'AES-GCM'),
    compression: normalizeCompressionMeta(rawMeta.compression),
    chunkSize,
    chunkCount,
    noncePrefix: rawMeta.noncePrefix,
    originalSize,
    mimeType: String(rawMeta.mimeType || 'application/octet-stream'),
  }
}

export function buildEncryptedChunkPlan(meta) {
  const chunks = []
  let offset = 0
  const variableEncryptedSizes = usesChunkCompression(meta)

  for (let chunkIndex = 0; chunkIndex < meta.chunkCount; chunkIndex += 1) {
    const start = chunkIndex * meta.chunkSize
    const plainSize = Math.min(meta.originalSize - start, meta.chunkSize)
    const encryptedSize = variableEncryptedSizes ? null : plainSize + 16

    chunks.push({ chunkIndex, encryptedSize, plainSize, variableEncryptedSize: variableEncryptedSizes })
    offset += encryptedSize || 0
  }

  return {
    chunks,
    totalEncryptedSize: variableEncryptedSizes ? null : offset,
    totalPlainSize: meta.originalSize,
    variableEncryptedSizes,
  }
}

function guessMimeType(filename) {
  const ext = path.extname(filename).toLowerCase()
  const map = {
    '.png': 'image/png',
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
    '.gif': 'image/gif',
    '.webp': 'image/webp',
    '.pdf': 'application/pdf',
    '.zip': 'application/zip',
    '.txt': 'text/plain',
    '.json': 'application/json',
    '.iso': 'application/octet-stream',
  }
  return map[ext] || 'application/octet-stream'
}

File: python/src/bytifi/crypto.py
from __future__ import annotations

import gzip
import math
import os
import secrets
from dataclasses import dataclass
from pathlib import Path

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from .base64url import from_base64url, to_base64url

ENCRYPTED_PART_SIZE = 32 * 1024 * 1024
PLAIN_CHUNK_SIZE = ENCRYPTED_PART_SIZE - 16
MULTIPART_THRESHOLD_BYTES = 10 * 1024 * 1024

_MIME_TYPES = {
    ".png": "image/png",
    ".jpg": "image/jpeg",
    ".jpeg": "image/jpeg",
    ".gif": "image/gif",
    ".webp": "image/webp",
    ".pdf": "application/pdf",
    ".zip": "application/zip",
    ".txt": "text/plain",
    ".json": "application/json",
    ".iso": "application/octet-stream",
}


def build_chunk_iv(nonce_prefix: bytes, chunk_index: int) -> bytes:
    iv = bytearray(12)
    iv[0:8] = nonce_prefix[0:8]
    iv[8:12] = chunk_index.to_bytes(4, "big")
    return bytes(iv)


def resolve_compression_mode(original_size: int = 0) -> str:
    if original_size > MULTIPART_THRESHOLD_BYTES:
        return "none"
    return "gzip"


def uses_chunk_compression(meta: dict) -> bool:
    compression = meta.get("compression") or {}
    return compression.get("algorithm") == "gzip"


def normalize_compression_meta(raw_compression: dict | None) -> dict:
    if not raw_compression or not isinstance(raw_compression, dict):
        return {"algorithm": "none", "scope": "chunk"}

    algorithm = str(raw_compression.get("algorithm", "none")).lower()
    if algorithm == "gzip":
        return {"algorithm": "gzip", "scope": "chunk"}
    return {"algorithm": "none", "scope": "chunk"}


def compress_plain_chunk(plain_chunk: bytes) -> bytes:
    return gzip.compress(plain_chunk)


def decompress_plain_chunk(compressed_chunk: bytes) -> bytes:
    return gzip.decompress(compressed_chunk)


def encrypt_chunk(payload_chunk: bytes, key: bytes, nonce_prefix: bytes, chunk_index: int) -> bytes:
    iv = build_chunk_iv(nonce_prefix, chunk_index)
    aesgcm = AESGCM(key)
    encrypted = aesgcm.encrypt(iv, payload_chunk, None)
    return encrypted


def decrypt_chunk(encrypted_chunk: bytes, key: bytes, nonce_prefix: bytes, chunk_index: int) -> bytes:
    if len(encrypted_chunk) < 16:
        raise ValueError("Encrypted chunk is too small.")
    iv = build_chunk_iv(nonce_prefix, chunk_index)
    aesgcm = AESGCM(key)
    return aesgcm.decrypt(iv, encrypted_chunk, None)


def decrypt_plain_chunk_from_encrypted(
    encrypted_chunk: bytes,
    key: bytes,
    nonce_prefix: bytes,
    chunk_index: int,
    meta: dict,
) -> bytes:
    payload = decrypt_chunk(encrypted_chunk, key, nonce_prefix, chunk_index)
    if not uses_chunk_compression(meta):
        return payload
    return decompress_plain_chunk(payload)


def build_client_encryption_meta(
    *,
    plain_chunk_size: int,
    chunk_count: int,
    nonce_prefix: bytes,
    original_size: int,
    mime_type: str,
    compression: str = "gzip",
) -> dict:
    compression_meta = (
        {"algorithm": "gzip", "scope": "chunk"}
        if compression == "gzip"
        else {"algorithm": "none", "scope": "chunk"}
    )
    return {
        "version": 2 if compression == "gzip" else 1,
        "algorithm": "AES-GCM",
        "compression": compression_meta,
        "chunkSize": plain_chunk_size,
        "chunkCount": chunk_count,
        "noncePrefix": to_base64url(nonce_prefix),
        "originalSize": original_size,
        "mimeType": mime_type or "application/octet-stream",
    }


def calculate_encrypted_size(original_size: int, plain_chunk_size: int = PLAIN_CHUNK_SIZE) -> tuple[int, int]:
    chunk_count = max(1, math.ceil(original_size / plain_chunk_size))
    encrypted_size = 0
    for chunk_index in range(chunk_count):
        start = chunk_index * plain_chunk_size
        plain_size = min(original_size - start, plain_chunk_size)
        encrypted_size += plain_size + 16
    return chunk_count, encrypted_size


@dataclass
class EncryptionContext:
    token: str
    token_bytes: bytes
    nonce_prefix: bytes
    meta: dict
    chunk_count: int
    encrypted_size: int
    original_name: str
    mime_type: str
    original_size: int
    plain_chunk_size: int
    compression: str


def create_encryption_context(
    *,
    original_size: int,
    original_name: str = "upload",
    mime_type: str = "application/octet-stream",
    token_bytes: bytes | None = None,
    nonce_prefix: bytes | None = None,
    plain_chunk_size: int = PLAIN_CHUNK_SIZE,
) -> EncryptionContext:
    token_bytes = token_bytes or secrets.token_bytes(32)
    nonce_prefix = nonce_prefix or secrets.token_bytes(8)

    if len(token_bytes) != 32:
        raise ValueError("Encryption token must be 32 bytes.")
    if len(nonce_prefix) != 8:
        raise ValueError("Nonce prefix must be 8 bytes.")

    compression = resolve_compression_mode(original_size)
    chunk_count, encrypted_size = calculate_encrypted_size(original_size, plain_chunk_size)
    meta = build_client_encryption_meta(
        plain_chunk_size=plain_chunk_size,
        chunk_count=chunk_count,
        nonce_prefix=nonce_prefix,
        original_size=original_size,
        mime_type=mime_type,
        compression=compression,
    )

    return EncryptionContext(
        token=to_base64url(token_bytes),
        token_bytes=token_bytes,
        nonce_prefix=nonce_prefix,
        meta=meta,
        chunk_count=chunk_count,
        encrypted_size=encrypted_size,
        original_name=original_name,
        mime_type=mime_type,
        original_size=original_size,
        plain_chunk_size=plain_chunk_size,
        compression=compression,
    )


def guess_mime_type(filename: str) -> str:
    ext = Path(filename).suffix.lower()
    return _MIME_TYPES.get(ext, "application/octet-stream")


def resolve_upload_file(file_path: str | os.PathLike, *, mime_type: str | None = None) -> tuple[Path, EncryptionContext]:
    absolute_path = Path(file_path).resolve()
    if not absolute_path.is_file():
        raise ValueError("Upload path must be a file.")

    original_name = absolute_path.name
    resolved_mime = mime_type or guess_mime_type(original_name)
    context = create_encryption_context(
        original_size=absolute_path.stat().st_size,
        original_name=original_name,
        mime_type=resolved_mime,
    )
    return absolute_path, context


def read_plain_chunk(file_path: Path, chunk_index: int, original_size: int, plain_chunk_size: int) -> bytes:
    start = chunk_index * plain_chunk_size
    length = min(plain_chunk_size, original_size - start)
    with file_path.open("rb") as handle:
        handle.seek(start)
        data = handle.read(length)
    if len(data) != length:
        raise ValueError("File ended before the upload was complete.")
    return data


def encrypt_chunk_from_file(file_path: Path, chunk_index: int, context: EncryptionContext) -> bytes:
    plain_chunk = read_plain_chunk(file_path, chunk_index, context.original_size, context.plain_chunk_size)
    payload = compress_plain_chunk(plain_chunk) if context.compression == "gzip" else plain_chunk
    return encrypt_chunk(payload, context.token_bytes, context.nonce_prefix, chunk_index)


def import_token(token: str) -> bytes:
    token_bytes = from_base64url(token)
    if len(token_bytes) != 32:
        raise ValueError("Invalid encryption token length.")
    return token_bytes


def normalize_client_encryption_meta(raw_meta: dict | None) -> dict | None:
    if not raw_meta or not isinstance(raw_meta, dict):
        return None

    chunk_size = int(raw_meta.get("chunkSize", 0))
    chunk_count = int(raw_meta.get("chunkCount", 0))
    original_size = int(raw_meta.get("originalSize", -1))
    nonce_prefix = raw_meta.get("noncePrefix")

    if chunk_size <= 0 or chunk_count <= 0 or original_size < 0:
        return None
    if not isinstance(nonce_prefix, str) or not nonce_prefix:
        return None

    return {
        "version": int(raw_meta.get("version", 1)),
        "algorithm": str(raw_meta.get("algorithm", "AES-GCM")),
        "compression": normalize_compression_meta(raw_meta.get("compression")),
        "chunkSize": chunk_size,
        "chunkCount": chunk_count,
        "noncePrefix": nonce_prefix,
        "originalSize": original_size,
        "mimeType": str(raw_meta.get("mimeType", "application/octet-stream")),
    }


def build_encrypted_chunk_plan(meta: dict) -> dict:
    chunks = []
    offset = 0
    variable = uses_chunk_compression(meta)

    for chunk_index in range(meta["chunkCount"]):
        start = chunk_index * meta["chunkSize"]
        plain_size = min(meta["originalSize"] - start, meta["chunkSize"])
        encrypted_size = None if variable else plain_size + 16
        chunks.append(
            {
                "chunkIndex": chunk_index,
                "encryptedSize": encrypted_size,
                "plainSize": plain_size,
                "variableEncryptedSize": variable,
            }
        )
        offset += encrypted_size or 0

    return {
        "chunks": chunks,
        "totalEncryptedSize": None if variable else offset,
        "totalPlainSize": meta["originalSize"],
        "variableEncryptedSizes": variable,
    }
